anchor-service/src/anchor_batch.rs

```diff
@@ -14,31 +21,121 @@ use crate::{
-        _batch_linger_time: Duration,
+        event_service: Arc<dyn Store>,
+        pool: SqlitePool,
+        node_id: NodeId,
+        anchor_interval: Duration,
+        anchor_batch_size: u64,
     ) -> Self {
         Self {
             tx_manager,
-            _event_service,
-            _batch_linger_time,
+            event_service,
+            high_water_mark_store: HighWaterMarkStore::new(pool),
+            node_id,
+            anchor_interval,
+            anchor_batch_size,
+        }
+    }
+
+    /// Run the AnchorService:
+    /// - Get anchor requests from the AnchorClient
+    /// - Deduplicate the anchor requests
+    /// - Anchor the batch using a Transaction Manager
+    /// - Store the TimeEvents using the AnchorClient
+    ///
+    /// This function will run indefinitely, or until the process is shutdown.
+    pub async fn run(&mut self, shutdown_signal: impl Future<Output = ()>) {
+        let mut interval = interval(self.anchor_interval);
+        let shutdown_signal = shutdown_signal.fuse();
+        pin_mut!(shutdown_signal);
+
+        info!("anchor service started");
+
+        loop {
+            let tick_future = interval.tick().fuse();
+            pin_mut!(tick_future);
+
+            match select(tick_future, &mut shutdown_signal).await {
+                Either::Left((_, _)) => {
+                    // Interval tick occurred
+                    if let Err(e) = self.process_next_batch().await {
+                        error!("Error processing batch: {:?}", e);
+                    }
+                }
+                Either::Right((_, _)) => {
+                    break;
+                }
+            }
+        }
+        info!("anchor service stopped");
+    }
+
+    async fn process_next_batch(&mut self) -> Result<()> {
+        // Pass the anchor requests through a deduplication step to avoid anchoring multiple Data Events from the same
+        // Stream.
+        let high_water_mark = self
+            .high_water_mark_store
+            .high_water_mark()
+            .await
+            .expect("error getting high water mark from database");
+        // Get the next batch of anchor requests
+        let anchor_requests: Vec<AnchorRequest> = match self
+            .event_service
+            .events_since_high_water_mark(
+                self.node_id,
+                high_water_mark,
+                self.anchor_batch_size as i64,
+            )
+            .await
+        {
+            Ok(requests) => IndexMap::<Cid, AnchorRequest>::from_iter(
+                requests.into_iter().map(|request| (request.id, request)),
+            )
+            .into_values()
+            .collect(),
+            Err(e) => {
+                return Err(e);
+            }
+        };
+        if anchor_requests.is_empty() {
+            info!("no requests to anchor");
+            return Ok(());
+        }
+        // Anchor the batch to the CAS. This may block for a long time.
```

Comment:
I don't think this will actually block the runtime, right? It is expected to take a long time, but it uses async IO so it will allow other uses of the system thread.. if that's not true, we should consider starting our own runtime or using system threads (but we should probably benchmark first)

Reply:
It will not block a system thread but the anchoring can't proceed until an anchor land.

Reply:
yeah, that's fair. it's a nit. you could modify the comment along those lines but no worries.